controllers/replication/secret_controller.go

```diff
@@ -455,23 +460,30 @@ func getPrefixedSecretName(prefix string, secretName string) string {
 	return fmt.Sprintf("%s%s", prefix, secretName)
 }
 
-func contains(s string, arr []string) bool {
-	for _, i := range arr {
-		if i == s {
-			return true
-		}
-	}
-	return false
-}
-
 func calculateTargetLabels(originalLabels map[string]string, target api.ReplicationTarget) map[string]string {
+	out := make(map[string]string)
+	for k, v := range originalLabels {
+		out[k] = v
+	}
 	for k, v := range target.AddLabels {
-		originalLabels[k] = v
+		out[k] = v
+	}
+	for _, key := range target.DropLabels {
+		delete(out, key)
 	}
-	for key := range originalLabels {
-		if contains(key, target.DropLabels) {
-			delete(originalLabels, key)
+	return out
+}
+
+func wouldBeInfinite(origin corev1.Secret, rsec api.ReplicatedSecret, target api.ReplicationTarget) bool {
+	computedLabels := calculateTargetLabels(origin.Labels, target)
+	for k, v := range rsec.Spec.Selector.MatchLabels {
+		if computedLabels[k] != v {
+			return false
 		}
 	}
-	return originalLabels
+	if (origin.Namespace == target.Namespace || target.Namespace == "") && target.K8sContextName == "" {
+		// This will still be infinite if the target has a non-empty k8scontext which points back to the origin cluster.
+		return true
+	}
```

Comment:
[question] Isn't `TargetPrefix != ""` also a precondition for infinite replication?

Reply:
True, but we call `wouldBeInfinite()` before checking the UID (in `requiresUpdate()`).

As a result, we log the warning `secret would be infinite, bailing` for things that would actually not be infinite, for example:
```yaml
apiVersion: v1
kind: Secret
metadata:
  namespace: k8ssandra-operator
  name: orig-secret
  labels:
    foo: bar
stringData:
---
apiVersion: replication.k8ssandra.io/v1alpha1
kind: ReplicatedSecret
metadata:
  namespace: k8ssandra-operator
  name: replicator
spec:
  selector:
    matchLabels:
      foo: bar
  replicationTargets:
    - namespace: k8ssandra-operator
```
That's not a big deal since that is a fringe case. But also it's a trivial fix.

Reply:
> True, but we call wouldBeInfinite() before checking the UID (in requiresUpdate()).

Perhaps I'm misunderstanding you, but I'm pretty sure that UID check will have two different UIDs if there is a target prefix, and you'll then land in this code:

```

	if srcHash, found := source.GetAnnotations()[coreapi.ResourceHashAnnotation]; found {
		// Get dest hash value
		destHash, destFound := dest.GetAnnotations()[coreapi.ResourceHashAnnotation]
		if !destFound {
			return true
		}

		if destSec, valid := dest.(*corev1.Secret); valid {
			hash := utils.DeepHashString(destSec.Data)
			if destHash != hash {
				// Destination data did not match destination hash
				return true
			}
		}

		return srcHash != destHash
```

All of that evaluates true because the names are different (so the hashes are as well, of course). At that point, you'll end up running:

```
logger.Info("Modifying secret in target cluster", "Secret", sec.Name, "TargetContext", target)
				syncSecrets(sec, fetchedSecret, target)
				copiedSecret := fetchedSecret.DeepCopy()
				copiedSecret.Name = getPrefixedSecretName(target.TargetPrefix, sec.Name)
				if err = remoteClient.Update(ctx, copiedSecret); err != nil {
					logger.Error(err, "Failed to sync target secret for matching payloads", "Secret", fetchedSecret.Name, "TargetContext", target)
					break TargetSecrets
				}
```

At that point you're actually in trouble, because you're trying to update a secret which doesn't (or shouldn't) exist - because it would have been infinite.

Basically this error message is there to let you know that you've misspecified your ReplicatedSecret and are risking infinite replication.